### DISCUSSION TITLE
Consider `defmulti` arity linting

### DISCUSSION THREAD

**Discussion by lread** [clj-kondo/clj-kondo#2002] [Ideas]
## Source of idea
On shadow-cljs slack chat, @p-himik was asking if shadow-cljs/cljs should help out with an error when calling a multimethod with the wrong arity. While he was talking about the compiler, I chimed in wondering if clj-kondo could help with a linter here.

@p-himik offered the following short and sweet snippet to show clj-kondo does not yet lint this type of thing:

```Shell
$ echo '(defmulti f identity) (f)' | clj-kondo --lint -
linting took 20ms, errors: 0, warnings: 0
```

## Mechanics of `defmulti` and `defmethod`
I started writing up a clj-kondo issue to propose a feature to lint for this use case, when I realized that I did not fully understand the mechanics of `defmulti` and `defmethod` arities. @borkdude and I started to explore.

## Question 1: Should arities of `defmethod`s and their `defmulti` dispatch fn all be equal?
Here's an example where arities do match:
```Clojure
user=> (defmulti foo (fn [x y] y))
#'user/foo
user=> (defmethod foo 2 [x y] {:x x :y y})
#object[clojure.lang.MultiFn 0x78830d9a "clojure.lang.MultiFn@78830d9a"]
user=> (foo 1 2)
{:x 1, :y 2}
```
But here's an example where the `defmulti` dispath fn is arity-1 and its `defmethod`s are arity-2.
```Clojure
user=> (defmulti foo :dispatch)
#'user/foo
user=> (defmethod foo :d1 [x y] {:x x :y y})
#object[clojure.lang.MultiFn 0x46c00568 "clojure.lang.MultiFn@46c00568"]
user=> (defmethod foo :d2 [a b] {:a a :b b})
#object[clojure.lang.MultiFn 0x46c00568 "clojure.lang.MultiFn@46c00568"]
user=> (foo {:dispatch :d1} "moar")
{:x {:dispatch :d1}, :y "moar"}
user=> (foo {:dispatch :d2} "moar")
{:a {:dispatch :d2}, :b "moar"}
```
Is this a normal/desirable thing to do?

## Question 2: Should `defmethod`s arities all be equal?
Although this is possible, it seems like a bad idea for `defmethod`s for a single `defmulti` to have different arities?

```Clojure
user=> (defmulti foo :dispatch)
#'user/foo
user=> (defmethod foo :d1 [x] {:x x})
#object[clojure.lang.MultiFn 0x2f879bab "clojure.lang.MultiFn@2f879bab"]
user=> (defmethod foo :d2 [x y] {:x x :y y})
#object[clojure.lang.MultiFn 0x2f879bab "clojure.lang.MultiFn@2f879bab"]
user=> (foo {:dispatch :d1})
{:x {:dispatch :d1}}
user=> (foo {:dispatch :d2} "moar")
{:x {:dispatch :d2}, :y "moar"}
```

## Question 3: Does multi-arity come into play?
Yes, it can, at least for `defmethod`s.

```Clojure
user=> (defmulti booya (fn [x & args] (:dispatch x)))
#'user/booya
user=> (defmethod booya :d1 ([x] x) ([x y] [x y]))
#object[clojure.lang.MultiFn 0x4690f583 "clojure.lang.MultiFn@4690f583"]
user=> (booya {:dispatch :d1})
{:dispatch :d1}
user=> (booya {:dispatch :d1} :moar)
[{:dispatch :d1} :moar]
```

## Question 4: Does ClojureScript behave the same as Clojure?
I expect linting rules should be the same, but the above experiments were all in Clojure, not ClojureScript.

**Comment by NoahTheDuke** [comment]
```
user=> (defmulti foo identity)
#'user/foo
user=> (defmethod foo :d1 [x y] {:x x :y y})
#object[clojure.lang.MultiFn 0x6ac97b84 "clojure.lang.MultiFn@6ac97b84"]
user=> (foo {:dispatch :d1} "moar")
Execution error (ArityException) at user/eval146 (REPL:1).
Wrong number of args (2) passed to: clojure.core/identity
```

Keywords have both 1-arity and 2-arity for "not-found", so I don't think this is so curious. All args are passed to both the dispatch fn and the resulting method.

**Reply by lread** [reply]
Ah... of course! Thanks @NoahTheDuke!

```Clojure
user=> (:foo [])
nil
user=> (:foo :foo)
nil
user=> (:foo :bar :baz)
:baz
```

**Comment by borkdude** [comment]
The dispatch function is called with the same amount of arguments as the `defmethod` function. Keywords may be called with 2 or 3 arguments. All arguments are first passed through the dispatch function and then through the defmethod function, is my understanding. If the dispatch function arities aren't compatible with the defmethod function, I think a clj-kondo warning should be emitted.

**Reply by lread** [reply]
A "doh" from me here on keywords, thanks!

**Reply by lread** [reply]
So my question 1 and question 2 do not apply.
